mlflow_client/backends.py

```diff
@@ -134,11 +140,51 @@ def end_run(self):
         pass
 
 
-class LocalBackend(Backend):
+class WithHooks(object):
+    """
+    Class decorator
+    Decorates a Backend to provide hooks on some methods
+    """
+    def __init__(self):
+        pass
+
+    def __call__(self, cls):
+        class BackendWithHooks(cls):
+            """
+            SubClass of the decorated Backend that implements Hooks
+            """
+            def __init__(self, *args, **kwargs):
+                # TODO: the problem is that hooks uri is not documented when creating a LocalBackend for example
+                uri = kwargs['hooks_uri'] if 'hooks_uri' in kwargs else None
+                self.hooks = resolve_hooks(uri)
+                super().__init__(*args, **kwargs)
+
+            def start_run(self, *args, **kwargs):
```

Comment:
I'm wondering if we can intercept any kind of method ?

Reply:
You can try this. We decorate each method we want to "hook" and pass the hook name as a decorator parameter.

```python
def log(s):
    def decorator(fn):
        def wrapper(self):
            print('[%s] before calling %s' % (s, fn.__name__))
            try:
                return fn(self)
            finally:
                print('[%s] done calling %s' % (s, fn.__name__))
        return wrapper
    return decorator

class C:
    @log("AAA")
    def printA(self):
        print("A")
    def printB(self):
        print("B")

C().printA()
C().printB()
```